content/articles/lru-cache-implementation-in-javascript/index.md

```diff
@@ -0,0 +1,135 @@
+
+We can optimize the shifting operation by eliminating it. To do so we use a double liked list. How?
+
+Still using the example 1,2,3,4. We needed to add 8. Since it wasn't in our list, what we did was remove the oldest element 4 and shift the entire list items. In a doubly-linked list, all you need to do is to remove the rear item and update the rear element to the previous element. Then add a new node and make it point to the second element and update the front. The entire operation is 0(1) time complexity. This will be clearly seen in the code example below.
+
+We can also optimize searching. How?
+
+Still using our previous example of cache size 4, with 1,2,3,4. searching for 8. We can optimize our search by using a hash table. With a Hash, you will have a key and a value pair. The Key in this case will be the actual values in our cache. The value will be the address of our value as shown in the diagram below.
+
+[Hash Table & Double Linked List Diagram](/ddl_hash.png/)
+
+To add 8 to our list, we will remove the rear element 4 in our list. 4 in this case is our Least recently accessed item. Update the rear element to point to the previous element 3. In our hash table, we will remove the value(address) that is associated with our key 4. Therefore, the value for our key 4 will be null. A new node will is created where we put the 8. The previous front gets updated to point to the node that has 8. The 8 will also have an address associated with it as well.
+
+The shifting, in this case, is done in 0(1).
+
+Before adding the 8, we need to search first and see if it is available in the cache. To search for the 8, we go to the hash table. In the Hash table, key 8 will not have a corresponding value attached to it. This means that it is not present in our Cache. This allows us to find out whether an item is in the Cache with just O(1) time.
+
+### Implementing LRU Cache in JavaScript.
+
+By visiting the below link, there is a leet code question to help us in the implementation of the LRU Cache. The code implementation below can guide us in the implementation.
+[List Recently Used Cache](https://leetcode.com/problems/lru-cache/)
+
+Below are the steps we will use to implement the LRU Cache class
+
+1.  We first initialize the LRU cache with a positive capacity.
+
+            ```js
+            var LRUCache = function(capacity) {
+                this.capacity = capacity;
+            this.map = new Map(); //this stores the entire array
+
+            //this is boundaries for double Linked List
+            this.head = {};
+            this.tail = {};
+            this.head.next = this.tail; //initialize your double Linked List
+            this.tail.prev = this.head;
+            };
+
+            ```

```

Comment:
Please use the correct javascript code indentation.

Also, **all** the code blocks should start at the beginning of the line.